ci: update ubuntu:20.04 builders to 22.04

cuviper opened the pull request:
This is mostly just maintenance to avoid bitrotting, but 22.04 also updates to cmake 3.22, so they don't need the manual builds from #113714 anymore.

cuviper committed:
ci: Update aarch64-gnu to ubuntu:22.04

cuviper committed:
ci: Update armhf-gnu to ubuntu:22.04

cuviper committed:
ci: Update dist-powerpc64le-linux to ubuntu:22.04

cuviper committed:
ci: Update dist-various-1 to ubuntu:22.04

cuviper committed:
ci: Update dist-various-2 to ubuntu:22.04

cuviper committed:
ci: Update dist-x86_64-musl to ubuntu:22.04

cuviper committed:
ci: Update dist-x86_64-netbsd to ubuntu:22.04

cuviper committed:
ci: Update test-various to ubuntu:22.04

cuviper committed:
ci: Update wasm32 to ubuntu:22.04

cuviper committed:
ci: Update x86_64-gnu-debug to ubuntu:22.04

cuviper committed:
ci: Update x86_64-gnu-nopt to ubuntu:22.04

cuviper committed:
ci: Update x86_64-gnu to ubuntu:22.04

wesleywiser commented on src/ci/docker/host-x86_64/dist-various-1/Dockerfile:
```diff
@@ -73,8 +74,8 @@ RUN env \
     env \
-    CC=arm-linux-gnueabihf-gcc CFLAGS="-march=armv7-a" \
-    CXX=arm-linux-gnueabihf-g++ CXXFLAGS="-march=armv7-a" \
+    CC=arm-linux-gnueabihf-gcc CFLAGS="-march=armv7-a -mfpu=vfp3" \
+    CXX=arm-linux-gnueabihf-g++ CXXFLAGS="-march=armv7-a -mfpu=vfp3" \
```
Just curious why this change was needed. Did gcc change the default here or something?

cuviper replied:
I'm not exactly sure what changed, but I do see that a lot of GCC tests explicitly specify this combination. Without it, the musl build fails:

```
+ cd musl-1.2.3
+ ./configure --enable-debug --disable-shared --prefix=/musl-armv7hf
checking for C compiler... arm-linux-gnueabihf-gcc
checking whether C compiler works... no; compiler output follows:
cc1: error: '-mfloat-abi=hard': selected architecture lacks an FPU
```

cuviper replied:
This may be a Ubuntu change -- in 20.04, `arm-linux-gnueabihf-gcc -v` reported configuration `--with-arch=armv7-a --with-fpu=vfpv3-d16`, and in 22.04 that's `--with-arch=armv7-a+fp` (w/o separate fpu).

cuviper replied:
In `man gcc` it says under armv7-a, "+fp The VFPv3 floating-point instructions, with 16 double-precision registers.  The extension +vfpv3-d16 can be used as an alias for this extension." So we should probably just match that and it will be consistent with what we had before.

cuviper committed:
ci: use `armv7-a+fp` for `armv7-unknown-linux_musleabihf`

This is consistent with what we had from `arm-linux-gnueabihf-gcc` in
the `ubuntu:20.04` base.

Mark-Simulacrum commented:
@bors r+ rollup=never

Nothing here looks too suspicious...

bors commented:
:pushpin: Commit 3cd3a91dbc50c71a3d51a32ac35a10869847c035 has been approved by `Mark-Simulacrum`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Mark-Simulacrum 3cd3a91dbc50c71a3d51a32ac35a10869847c035 -->
<!-- homu: {"type":"Approved","sha":"3cd3a91dbc50c71a3d51a32ac35a10869847c035","approver":"Mark-Simulacrum","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 3cd3a91dbc50c71a3d51a32ac35a10869847c035 with merge b6dd153fbcc06b3bc936aca800af4b8376c229e4...
<!-- homu: {"type":"BuildStarted","head_sha":"3cd3a91dbc50c71a3d51a32ac35a10869847c035","merge_sha":"b6dd153fbcc06b3bc936aca800af4b8376c229e4"} -->

Kobzol commented:
I *think* that the bump from 20.04 to 22.04 updates the glibc version. Do we care about this for some of the modified dist builders? Or is this only relevant for x64 Linux gnu?

cuviper commented:
Hmm, I suppose it is relevant to any `*-linux-gnu` that uses the Ubuntu cross compilers, but it doesn't matter as much for those that don't provide host tools. So `dist-powerpc64le-linux` is a host but not using the Ubuntu toolchain; `dist-various-1` and `-2` have a few `linux-gnu` that are non-host, using Ubuntu's toolchain.

Do we care about `libstd.so` glibc versions on those? (`libstd.rlib` should be fine.)

cuviper commented:
If we do care, maybe we could fork those `linux-gnu` targets into a `dist-various-3` or something, that explicitly holds back on `ubuntu:20.04`. Long term, it would be better to use crosstool-ng toolchains when we want to control glibc versions.

Mark-Simulacrum commented:
Can you check against https://doc.rust-lang.org/beta/rustc/platform-support.html? I think if we don't document a version we should be fine to bump as a nonevent, otherwise let's bump the documentation at least. (My quick scroll suggests it's not unlikely this doesn't change anything, but I'm not sure)

matthiaskrgr commented:
Hmm, `dist-x86_64-musl` is running for 2.5 hours already, hopefully that is just a one-off :sweat:

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/5694729785/job/15440106219)
Approved by: Mark-Simulacrum
Pushing b6dd153fbcc06b3bc936aca800af4b8376c229e4 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Mark-Simulacrum","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/5694729785/job/15440106219"},"merge_sha":"b6dd153fbcc06b3bc936aca800af4b8376c229e4"} -->

cuviper commented:
> Can you check against https://doc.rust-lang.org/beta/rustc/platform-support.html? I think if we don't document a version we should be fine to bump as a nonevent, otherwise let's bump the documentation at least.

The glibc versions under Tier 2 (without host) should be updated, but they're already wrong. They should have been updated for #81521 at least, and there's a deja vu comment about compat and dynamic linking... :sweat_smile: 

I'll send an update, including a footnote about that applicability.

cuviper commented:
> Hmm, `dist-x86_64-musl` is running for 2.5 hours already, hopefully that is just a one-off 😓

LLVM alone took over 1.5 hours in that build, but only 5 minutes in the next -- hooray for sccache!

rust-timer commented:
Finished benchmarking commit (b6dd153fbcc06b3bc936aca800af4b8376c229e4): [comparison URL](https://perf.rust-lang.org/compare.html?start=317ec04d18ab9c5d09a4d9b4418a12da8974b961&end=b6dd153fbcc06b3bc936aca800af4b8376c229e4&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=317ec04d18ab9c5d09a4d9b4418a12da8974b961&end=b6dd153fbcc06b3bc936aca800af4b8376c229e4&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -0.6% | [-0.8%, -0.3%] | 3     |
| All ❌✅ (primary)                 | -     | -              | 0     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=317ec04d18ab9c5d09a4d9b4418a12da8974b961&end=b6dd153fbcc06b3bc936aca800af4b8376c229e4&stat=max-rss)
This benchmark run did not return any relevant results for this metric.

### [Cycles](https://perf.rust-lang.org/compare.html?start=317ec04d18ab9c5d09a4d9b4418a12da8974b961&end=b6dd153fbcc06b3bc936aca800af4b8376c229e4&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 3.3% | [1.0%, 6.3%] | 8     |
| Regressions ❌ <br /> (secondary)  | 2.1% | [2.1%, 2.1%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 3.3% | [1.0%, 6.3%] | 8     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=317ec04d18ab9c5d09a4d9b4418a12da8974b961&end=b6dd153fbcc06b3bc936aca800af4b8376c229e4&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 650.027s -> 649.861s (-0.03%)
